### ISSUE TITLE
Questions Regarding SFT Training with TIR Data

### ISSUE THREAD

**Issue by chenzihan1** [NVIDIA/NeMo-Skills#483]
Hi NeMo-Skills Team,

I have some questions regarding the Supervised Fine-Tuning (SFT) process with the Tool-Integrated Reasoning (TIR) corpus provided in the project.

I've noticed that in the open-sourced TIR corpus, the tool call invocations (<tool_call>...</tool_call>), their execution results (e.g., output ...), and even system messages (e.g., system ...) appear to be interleaved with the model's reasoning steps, as illustrated by this example from the data:
```python
Let me write the Python code for this.
<tool_call>
import itertools

# Define the possible scores for each subject (0 to 10)
scores = range(11)

# Generate all possible combinations of four subjects
combinations = itertools.product(scores, repeat=4)

# Filter combinations where the total score is 21
valid_combinations = [combo for combo in combinations if sum(combo) == 21]

# The number of valid combinations is the answer
num_ways = len(valid_combinations)
num_ways
</tool_call>
```output
880
```system
You have run out of code executions! You can no longer write or execute code. Now you should continue solving the problem by relying on your mathematical reasoning and analytical skills.
/```
Think_content
```
This leads to a few points of clarification I'd appreciate your insights on:

Masking during SFT: During Supervised Fine-Tuning (SFT) with this corpus, are the tool execution results (e.g., the 880 within the output block) and the system messages masked out from the loss calculation? My understanding is that in a real-world scenario, these parts are provided by the environment/tool executor or are system-generated directives, not something the model itself should be trained to predict as part of its own output. Is my understanding incorrect here?

Origin of System Messages: Related to the above, are the system messages also intended to be generated by the model during inference, or are they external inputs to guide the model?

Special Handling for TIR SFT Data: For SFT, is there any special preprocessing or handling required for this TIR corpus beyond standard language modeling objectives, particularly concerning these interleaved tool calls, outputs, and system messages? Or should the model be trained to generate the entire sequence as presented?

Understanding these aspects of the SFT process would be very helpful for us as we prepare and fine-tune our models for effective TIR.

Thanks for your guidance.

Best regards!!!

**Comment by Kipok** [comment]
> are the tool execution results (e.g., the 880 within the output block) and the system messages masked out from the loss calculation?

No, we don't mask out the code execution results. We never really tried, so it could work better potentially, but we don't do it.

> are the system messages also intended to be generated by the model during inference, or are they external inputs to guide the model?

The system message after code blocks is not generated by the model. We explicitly add it over here https://github.com/NVIDIA/NeMo-Skills/blob/main/nemo_skills/code_execution/utils.py#L33

> Or should the model be trained to generate the entire sequence as presented?

That's what we do, no special handling. But we didn't really try to change that, so let us know if masking those out works better for you.

**Comment by Kipok** [comment]
Feel free to reopen if you have more questions.
